### PR TITLE
Add generation for all IOperation nodes

### PR DESCRIPTION
**Pull request by 333fred**
Almost all IOperation implementations in Microsoft.CodeAnalysis are now being generated. Exceptions are main NoneOperation, InvalidOperation, and the operations that involve dynamic, as we have a special base class they all inherit from internally that's used to implement some public API functionality. Unfortunately, as in the previous PR, I was unable to generate these new classes in-place. I've done some reordering of properties in the xml to make the constructors generate parameters in the same order as they did previously, but in a few cases that wasn't possible, and for those cases I've reordered parameters in other files. I've also standardized the names of the IConvertibleConversion properties, which caused a few renames in the code as well.

I've also added a verification step for the XML, to help prevent simple errors from being introduced. It will ensure that children have an explicit ordering if required, that operations follow naming conventions, that base types are correctly specified, and that summaries are appropriately included. @dotnet/roslyn-compiler please review.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler please review.

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -3504,14 +3504,14 @@ private void AddExceptionStore(ITypeSymbol exceptionType, IOperation exceptionDe
                 if (exceptionTarget != null)
                 {
                     AddStatement(new SimpleAssignmentOperation(
-                                         exceptionTarget,
                                          isRef: false,
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: formatting?

File: src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs [hunk 2]
```diff
@@ -4204,25 +4205,34 @@ IOperation getLoopControlVariableAssignment(IOperation current)
...
                         return new DeconstructionAssignmentOperation(VisitPreservingTupleOperations(operation.LoopControlVariable),
-                                                                      current, semanticModel: null,
-                                                                      operation.LoopControlVariable.Syntax, operation.LoopControlVariable.Type,
-                                                                      constantValue: default, isImplicit: true);
+                                                                     current, semanticModel: null,
+                                                                     operation.LoopControlVariable.Syntax, operation.LoopControlVariable.Type,
+                                                                     constantValue: default, isImplicit: true);
                     default:
-                        return new SimpleAssignmentOperation(Visit(operation.LoopControlVariable),
-                                                              isRef: false, // In C# this is an error case and VB doesn't support ref locals
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs]: seems like this comment was meant to remain attached to`isRef: false`?

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml [hunk 3]
```diff
@@ -77,7 +77,7 @@
       </Comments>
     </Property>
   </Node>
-  <Node Name="ISwitchOperation" Base="IOperation">
+  <Node Name="ISwitchOperation" Base="IOperation" ChildrenOrder="Value,Cases">
```
Comment (RikkiGibson) [inline src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: >ChildrenOrder [](start = 50, length = 13)

It seems like sometimes it's validated that all properties are present in ChildrenOrder, but properties like Locals are not included. What am I missing about the way the validation works?

**Comment by cston** [comment]
>                     ImmutableInterlocked.InterlockedCompareExchange(ref _lazyOperations, operations, default);

`ImmutableInterlocked.InterlockedInitialize()`

---
Refers to: src/Compilers/Core/Portable/Generated/Operations.Generated.cs:2808 in ddee7cd. [](commit_id = ddee7cd544756b1965e2a8216af9674dde8dc13d, deletion_comment = False)

**Comment by cston** [comment]
> using System;

Copyright header.

---
Refers to: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs:1 in ddee7cd. [](commit_id = ddee7cd544756b1965e2a8216af9674dde8dc13d, deletion_comment = False)

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs [hunk 4]
```diff
@@ -340,80 +348,202 @@ private void WriteClasses()
                 if (type.SkipClassGeneration) continue;
 
                 var allProps = GetAllProperties(type);
+                bool hasSkippedProperties = !GetAllProperties(type, includeSkipGenerationProperties: true).SequenceEqual(allProps);
                 var ioperationTypes = allProps.Where(p => IsIOperationType(p.Type)).ToList();
```
Comment (cston) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs]: >ioperationTypes [](start = 20, length = 15)

Consider naming `ioperationProps`.

File: src/Compilers/Core/Portable/Operations/OperationInterfaces.xml (hunk 3 above)
Comment (333fred) [reply src/Compilers/Core/Portable/Operations/OperationInterfaces.xml]: Children are only `IOperation` or arrays of `IOperation`.

---
In reply to: [315387892](https://github.com/dotnet/roslyn/pull/38056#discussion_r315387892) [](ancestors = 315387892)

**Comment by 333fred** [comment]
> using System;

I wanted to break this up to avoid putting too much in one file, I'll add a copyright header.

---
In reply to: [522740001](https://github.com/dotnet/roslyn/pull/38056#issuecomment-522740001) [](ancestors = 522740001)

---
Refers to: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs:1 in ddee7cd. [](commit_id = ddee7cd544756b1965e2a8216af9674dde8dc13d, deletion_comment = False)

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+                    {
+                        Console.WriteLine($"{abstractNode.Name} does not have correctly formatted comments, please ensure that there is a <summary> block for the type.");
+                        error = true;
+                    }
+
+                    foreach (var prop in abstractNode.Properties)
+                    {
+                        if (prop.Comments?.Elements?[0].Name != "summary" && !prop.IsInternal)
+                        {
+                            Console.WriteLine($"{abstractNode.Name}.{prop.Name} does not have correctly formatted comments, please ensure that there is a <summary> block for the type.");
```
Comment (cston) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs]: >type [](start = 178, length = 4)

"property"?

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+            {
+                if (!abstractNode.Name.StartsWith("I"))
+                {
+                    Console.WriteLine($"All IOperation node names must start with I. {abstractNode.Name} does not.");
+                    error = true;
+                }
+
+                if (!abstractNode.Name.EndsWith("Operation"))
+                {
+                    Console.WriteLine($"All IOperation node names must end with Operation. {abstractNode.Name} does not.");
```
Comment (333fred) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs]: Add error = true;

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+                            error = true;
+                        }
+                    }
+                    continue;
+                }
+
+
+                if (node.ChildrenOrder is null)
+                {
+                    Console.WriteLine($"{node.Name} has more than 1 IOperation properties and must declare an explicit ordering with the ChildrenOrder attribute.");
```
Comment (cston) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs]: >properties [](start = 79, length = 10)

"property"?

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+                    Console.WriteLine($"Properties: {string.Join(", ", properties)}");
+                    error = true;
+                    continue;
+                }
+
+                var childrenOrdered = node.ChildrenOrder.Split(",", StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()).ToList();
```
Comment (cston) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs]: >.Split(",", StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()) [](start = 56, length = 72)

Perhaps extract a helper method.

**Comment by 333fred** [comment]
@rikkigibson @cston addressed feedback.

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs [hunk 9]
```diff
@@ -340,80 +348,202 @@ private void WriteClasses()
                 if (type.SkipClassGeneration) continue;
 
                 var allProps = GetAllProperties(type);
-                var ioperationTypes = allProps.Where(p => IsIOperationType(p.Type)).ToList();
-                if (ioperationTypes.Count == 0)
+                bool hasSkippedProperties = !GetAllProperties(type, includeSkipGenerationProperties: true).SequenceEqual(allProps);
+                var ioperationProperties = allProps.Where(p => IsIOperationType(p.Type)).ToList();
+                var publicIOperationTypes = ioperationProperties.Where(p => !p.IsInternal).ToList();
```
Comment (cston) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.cs]: >publicIOperationTypes [](start = 20, length = 21)

`publicIOperationProperties`?

File: src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+                if (!(abstractNode is Node node)) continue;
+                if (node.SkipChildrenGeneration || node.SkipClassGeneration) continue;
+
+                var properties = GetAllGeneratedIOperationProperties(node).Where(p => !p.IsInternal).Select(p => p.Name).ToList();
+
+                if (properties.Count < 2)
+                {
+                    if (node.ChildrenOrder is string order)
+                    {
+                        var splitOrder = order.Split(",", StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()).ToList();
```
Comment (cston) [inline src/Tools/Source/CompilerGeneratorTools/Source/IOperationGenerator/IOperationClassWriter.Verifier.cs]: >order.Split(",", StringSplitOptions.RemoveEmptyEntries).Select(s => s.Trim()).ToList(); [](start = 41, length = 87)

`GetPropertyOrder(order)`

**Pull request merged**